src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml

```diff
@@ -1,243 +1,232 @@
+    we want the generator to call itsef.
+ *)
 
 let rec longident_to_str = function
   | Lident s -> s
   | Ldot (lg, s) -> Printf.sprintf "%s.%s" (longident_to_str lg) s
   | Lapply (lg1, lg2) ->
       Printf.sprintf "%s %s" (longident_to_str lg1) (longident_to_str lg2)
 
-let rec is_rec_typ typ_name = function
+let rec is_rec_typ typ_names = function
   | { ptyp_desc = Ptyp_constr ({ txt = x; _ }, _); _ } ->
-      longident_to_str x = typ_name
-  | { ptyp_desc = Ptyp_tuple xs; _ } -> List.exists (is_rec_typ typ_name) xs
+     List.exists (fun typ_name -> longident_to_str x = typ_name) typ_names
+  | { ptyp_desc = Ptyp_tuple xs; _ } -> List.exists (is_rec_typ typ_names) xs
+  | { ptyp_desc = Ptyp_variant (rws, _, _); _ } ->
+     List.exists (fun rw ->
+         match rw.prf_desc with
+         | Rtag (lab, _, cts) ->
+            List.exists (fun typ_name -> lab.txt = typ_name) typ_names ||
+              List.exists (is_rec_typ typ_names) cts
+         | Rinherit ct -> is_rec_typ typ_names ct) rws
   | _ -> false
 
-let name s =
-  let prefix = "gen" in
-  match s with "t" -> prefix | s -> prefix ^ "_" ^ s
+let is_rec_constr_decl typ_names cd =
+  match cd.pcd_args with
+  | Pcstr_tuple cts -> List.exists (is_rec_typ typ_names) cts
+  | _ -> false
 
-let pat ~loc s =
-  let (module A) = Ast_builder.make loc in
-  let s = name s in
-  A.pvar s
+(** [is_rec_type_decl typ_names typ] looks for elements of [typ_names]
+    recursively in [typ]. *)
+let is_rec_type_decl typ_names typ =
+  let in_type_kind =
+    match typ.ptype_kind with
+    | Ptype_variant cstrs -> List.exists (is_rec_constr_decl typ_names) cstrs
+    | _ -> false
+  in
+  let in_manifest =
+    match typ.ptype_manifest with
+    | Some x -> is_rec_typ typ_names x
+    | None -> false
+  in
+  in_type_kind || in_manifest
+
+(** [env] contains the list of recursive types during the derivation
+
+    i.e:
+    - contains one element maximum if its a single type declaration *)
+type env = {
+    rec_types : string list;
+    curr_types : string list;
+    curr_type : string;
+  }
 
-let gen ~loc ?(env = TypeGen.empty) lg =
+(** [add_if_rec env tys ty x] adds x to [env.rec_types] if typ is recursive *)
+let add_if_rec env typ x =
+  if is_rec_type_decl env.curr_types typ then
+    { env with rec_types = x :: env.rec_types}
+  else env
+
+(** {2. Generator constructors} *)
+
+(** [gen_longident lg args] creates a generator using [lg].
+
+    The longident can either be a:
+    - Lident s: We transform to gen_s (or gen if s = "t")
+    - Ldot (lg, s): We transform to qualified generator (e.g. B.gen)
+*)
+let gen_longident ~loc ~env lg args =
   let (module A) = Ast_builder.make loc in
   match lg with
   | Lident s ->
-      Option.value ~default:(name s |> A.evar) @@ TypeGen.find_opt s env
-  | Ldot (lg, s) -> A.(pexp_ident (Located.mk @@ Ldot (lg, name s)))
+     if List.mem s env.rec_types then
```

Comment:
In this case I think the code actually reads clearer with a helper function over environments:
```ocaml
if is_rec env s then
```
There are a bunch of other ad-hoc `is_rec` at the bottom which could be eliminated and instead reuse such a helper.

Ideally `type env` and these helpers could be collected in an `Env` module 😀


However there's also something weird about having parallel `is_rec*`
- for `string`s and
- for `core_type` 🤔

Reply:
I introduced a module `Env` with the type and `is_rec`. I'm do not think I can add other helpers functions.

I believe having `is_rec` in `Env` reduces the weirdness:
* If you do `Env.is_rec` you ask if the type is recursive based on the previously computed environment
* If you do `is_rec_XXX` you ask to compute whether the type is recursive based on the known environment